### PR TITLE
FM2-83 : Create appropriate search functionality for allergy intolerance resource

### PR DESCRIPTION
**Pull request by jecihjoy**
## Description of what I changed
Implemented search by patient reference, allergy category, allergy substance, severity, manifestation/reaction and clinical status

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-83

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**.

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirAllergyIntoleranceServiceImpl.java [hunk 1]
```diff
@@ -11,8 +11,16 @@
...
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+import ca.uhn.fhir.rest.param.ReferenceParam;
+import ca.uhn.fhir.rest.param.StringOrListParam;
+import ca.uhn.fhir.rest.param.TokenOrListParam;
 import lombok.AccessLevel;
 import lombok.Setter;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
```
Comment (corneliouzbett) [inline api/src/main/java/org/openmrs/module/fhir2/api/impl/FhirAllergyIntoleranceServiceImpl.java]: Imported but never used

File: omod/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceFhirResourceProvider.java [hunk 2]
```diff
@@ -13,15 +13,25 @@
...
+import ca.uhn.fhir.rest.annotation.Search;
+import ca.uhn.fhir.rest.param.ReferenceParam;
+import ca.uhn.fhir.rest.param.StringOrListParam;
+import ca.uhn.fhir.rest.param.TokenOrListParam;
 import ca.uhn.fhir.rest.server.IResourceProvider;
 import ca.uhn.fhir.rest.server.exceptions.ResourceNotFoundException;
 import lombok.AccessLevel;
 import lombok.Setter;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
```
Comment (corneliouzbett) [inline omod/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceFhirResourceProvider.java]: Ditto

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java [hunk 3]
```diff
@@ -10,28 +10,181 @@
...
+		if (statusParam != null && !statusParam.getValuesAsQueryTokens().isEmpty()) {
+			switch (statusParam.getValuesAsQueryTokens().get(0).getValue()) {
+				case "active":
+					return new TokenOrListParam().add("false");
+				case "inactive":
+					return new TokenOrListParam().add("true");
+			}
+		}
+		return null;
+	}
```
Comment (corneliouzbett) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java]: Why do you pass statusParam as `TokenOrListParam`

File: api/src/main/java/org/openmrs/module/fhir2/api/FhirAllergyIntoleranceService.java [hunk 4]
```diff
@@ -11,9 +11,18 @@
...
+import ca.uhn.fhir.rest.param.TokenOrListParam;
 import org.hl7.fhir.r4.model.AllergyIntolerance;
 
 public interface FhirAllergyIntoleranceService {
 	
 	AllergyIntolerance getAllergyIntoleranceByUuid(@NotNull String uuid);
+	
+	Collection<AllergyIntolerance> searchForAllergies(ReferenceParam patientReference, StringOrListParam category,
+	        TokenOrListParam allergen, TokenOrListParam severity, TokenOrListParam manifestationCode,
+	        StringOrListParam clinicalStatus);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/FhirAllergyIntoleranceService.java]: `clinicalStatus` should be a `TokenOrListParam`. The definitions for the values taken by these parameters can be found [here](https://www.hl7.org/fhir/allergyintolerance.html#search). Note that the difference between a `TokenParam` and a `StringParam` is that the `TokenParam` can contain a `system` value indicating what value-set the value it encodes represents.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/FhirAllergyIntoleranceDao.java [hunk 5]
```diff
@@ -11,9 +11,20 @@
...
+import org.hl7.fhir.r4.model.Patient;
 import org.openmrs.Allergy;
 
 public interface FhirAllergyIntoleranceDao {
 	
 	Allergy getAllergyIntoleranceByUuid(@NotNull String uuid);
+	
+	Collection<Allergy> searchForAllergies(@OptionalParam(name = Patient.SP_IDENTIFIER) ReferenceParam patientReference,
+	        StringOrListParam category, TokenOrListParam allergen, TokenOrListParam severity,
+	        TokenOrListParam manifestationCode, StringOrListParam clinicalStatus);
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/FhirAllergyIntoleranceDao.java]: As above for `clinicalStatus`

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/FhirAllergyIntoleranceDao.java [hunk 6]
```diff
@@ -11,9 +11,20 @@
...
+import ca.uhn.fhir.rest.param.StringOrListParam;
+import ca.uhn.fhir.rest.param.TokenOrListParam;
+import org.hl7.fhir.r4.model.Patient;
 import org.openmrs.Allergy;
 
 public interface FhirAllergyIntoleranceDao {
 	
 	Allergy getAllergyIntoleranceByUuid(@NotNull String uuid);
+	
+	Collection<Allergy> searchForAllergies(@OptionalParam(name = Patient.SP_IDENTIFIER) ReferenceParam patientReference,
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/FhirAllergyIntoleranceDao.java]: `@OptionalParam` does not belong here, but only on the `ResourceProvider`.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java [hunk 7]
```diff
@@ -10,28 +10,181 @@
...
+		}
+		
+		criteria.createAlias("allergen.codedAllergen", "c");
+		
+		handleOrListParamBySystem(allergen, (system, tokens) -> {
+			if (system.isEmpty()) {
+				return Optional.of(in("c.uuid", tokensToList(tokens)));
+			} else {
+				if (!containsAlias(criteria, "cs")) {
+					criteria.createAlias("c.uuid", "cs");
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java]: This seems to be creating a join on the `codedAllergen.uuid` field, which is not going to work.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java [hunk 8]
```diff
@@ -10,28 +10,181 @@
...
+		}
+		criteria.createAlias("reactions", "r");
+		criteria.createAlias("r.reaction", "c");
+		
+		handleOrListParamBySystem(manifestation, (system, tokens) -> {
+			if (system.isEmpty()) {
+				return Optional.of(in("c.uuid", tokensToList(tokens)));
+			} else {
+				if (!containsAlias(criteria, "rs")) {
+					criteria.createAlias("c.uuid", "rs");
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java]: This seems to be creating a join on the `r.reaction.uuid` field, which is not going to work.

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java [hunk 9]
```diff
@@ -10,28 +10,181 @@
...
+				return Optional.of(or(eq("cs.name", system), in("c.uuid", tokensToList(tokens))));
+			}
+		}).ifPresent(criteria::add);
+	}
+	
+	private Optional<Criterion> handleSeverity(Criteria criteria, TokenOrListParam severityParam) {
+		if (severityParam == null) {
+			return Optional.empty();
+		}
+		criteria.createAlias("severity", "c");
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java]: Please add a new line here

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java [hunk 10]
```diff
@@ -10,28 +10,181 @@
...
+		
+		return handleOrListParam(severityParam, token -> {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/FhirAllergyIntoleranceDaoImpl.java]: This is a proper way to implement this; however, it is probably preferable to cache the values of the calls to `getConceptId(globalPropertyService.getGlobalProperty(...))` outside of the lambda expression. something like this:

```java
final int mildConceptId = getConceptId(globalPropertyService.getGlobalProperty(FhirConstants.GLOBAL_PROPERTY_MILD));
...
return handleOrListParam(severityParam, token -> {
    try {
        AllergyIntolerance.AllergyIntoleranceSeverity severity = AllergyIntolerance.AllergyIntoleranceSeverity
            .fromCode(token.getValue());
        
        switch(servity) {
             case MILD:
                 return Optional.of(eq("c.conceptId", mildConceptId));
             ...
        }
    }
    ...
});
```

Thinking about this some more, we should probably create a method on the `globalPropertyService` called `getGlobalProperties(String... properties)` that looks something like this:

```java
public Collection<String> getGlobalProperties(String... properties) {
    List<GlobalProperty> globalProperties = (GlobalProperty) sessionFactory.getCurrentSession()
        .createCriteria(GlobalProperty.class).add(Restrictions.in("property", properties)).list();
    return globalProperties.stream().filter(p -> p != null).map(GlobalProperty::getPropertyValue).collect(Collectors.toList());
}
```

Then we can use this:

```java
final int[] severityConceptUuids = globalPropertyService.getGlobalProperties(FhirConstants.GLOBAL_PROPERTY_MILD, FhirConstants.GLOBAL_PROPERTY_MODERATE, FhirConstants.GLOBAL_PROPERTY_SEVERE, FhirConstants.GLOBAL_PROPERTY_OTHER);
...
return handleOrListParam(severityParam, token -> {
    try {
        AllergyIntolerance.AllergyIntoleranceSeverity severity = AllergyIntolerance.AllergyIntoleranceSeverity
            .fromCode(token.getValue());
        
        switch(servity) {
             case MILD:
                 return Optional.of(eq("c.uuid", severityConceptIds[0]));
             ...
        }
    }
    ...
});
```

File: omod/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceFhirResourceProvider.java [hunk 11]
```diff
@@ -47,4 +55,19 @@ public AllergyIntolerance getAllergyIntoleranceByUuid(@IdParam @NotNull IdType i
...
 		return allergy;
 	}
+	
+	@Search
+	@SuppressWarnings("unused")
+	public Bundle searchForAllergies(
+	        @OptionalParam(name = AllergyIntolerance.SP_PATIENT, chainWhitelist = { "", Patient.SP_IDENTIFIER,
+	                Patient.SP_GIVEN, Patient.SP_FAMILY,
+	                Patient.SP_NAME }, targetTypes = Patient.class) ReferenceParam patientReference,
+	        @OptionalParam(name = AllergyIntolerance.SP_CATEGORY) StringOrListParam category,
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceFhirResourceProvider.java]: `TokenOrListParam`

File: omod/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceFhirResourceProvider.java [hunk 12]
```diff
@@ -47,4 +55,19 @@ public AllergyIntolerance getAllergyIntoleranceByUuid(@IdParam @NotNull IdType i
...
+	        @OptionalParam(name = AllergyIntolerance.SP_CODE) TokenOrListParam allergen,
+	        @OptionalParam(name = AllergyIntolerance.SP_SEVERITY) TokenOrListParam severity,
+	        @OptionalParam(name = AllergyIntolerance.SP_MANIFESTATION) TokenOrListParam manifestationCode,
+	        @OptionalParam(name = AllergyIntolerance.SP_CLINICAL_STATUS) StringOrListParam clinicalStatus) {
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/providers/AllergyIntoleranceFhirResourceProvider.java]: `TokenOrListParam`

**Pull request merged**